simple_network_sim/inference.py

```diff
@@ -0,0 +1,708 @@
+        self.initial_infections_stddev = parameters["initial_infections_stddev"]
+        self.initial_infections_stddev_min = parameters["initial_infections_stddev_min"]
+        self.initial_infections_kernel_sigma = parameters["initial_infections_kernel_sigma"]
+
+        assert self.n_smc_steps > 0
+        assert self.n_particles > 0
+        assert self.infection_probability_shape > 0.
+        assert self.infection_probability_kernel_sigma > 0.
+        assert self.initial_infections_stddev > 0.
+        assert self.initial_infections_stddev_min > 0.
+        assert self.initial_infections_kernel_sigma > 0.
+
+        self.compartment_transition_table = compartment_transition_table
+        self.population_table = population_table
+        self.commutes_table = commutes_table
+        self.mixing_matrix_table = mixing_matrix_table
+        self.infection_probability = infection_probability
+        self.initial_infections = initial_infections
+        self.infectious_states = infectious_states
+        self.trials = trials
+        self.start_end_date = start_end_date
+        self.movement_multipliers_table = movement_multipliers_table
+        self.stochastic_mode = stochastic_mode
+        self.random_seed = random_seed
+        self.random_state = np.random.default_rng(loaders.readRandomSeed(random_seed))
+
+        assert trials.at[0, "Value"] == 1
+
+        self.threshold = np.inf
+        self.fit_statistics: Dict[int] = {}
+
+    def fit(self) -> Tuple[List[Particle], List[float], List[float]]:
+        """ Performs ABC-SMC iterative procedure for finding posterior distributions
+        of model parameters given priors and data. In brief, iteratively samples
+        particles, keeping at each round only ones with good fitting criteria.
+
+        References:
+        https://royalsocietypublishing.org/doi/pdf/10.1098/rsif.2008.0172
+        https://en.wikipedia.org/wiki/Approximate_Bayesian_computation
+
+        :return: List of particles selected at the end of the procedure, and weights
+        associated
+        """
+        prev_particles: List[Particle] = []
+        prev_weights: List[float] = []
+        distances: List[float] = []
+
+        for smc_step in range(self.n_smc_steps):
+            logger.info("SMC step %d/%d", smc_step + 1, self.n_smc_steps)
+            prev_particles, prev_weights, distances = self.sample_particles(smc_step, prev_particles, prev_weights)
+            self.update_threshold(distances)
+
+        return prev_particles, prev_weights, distances
+
+    def update_threshold(self, distances: List[float]):
+        """ Updates threshold using distances found on previous round.
+        The median is used as a criterion for next round.
+
+        :param distances: List of distances found for particles in previous round
+        :return: New threshold to use for next round, median of distances from previous round
+        """
+        self.threshold = np.percentile(distances, 50)
```

Comment:
Is this method of decaying the threshold standard? I imagine so, just curious.

Reply:
I think so, from what I saw there are a two choices:
- Manually input a decreasing list. The decrease could be e.g. linear (threshold=threshold_prev-1), or multiplicative (threshold=threshold*0.8). This works well when we know which values to expect and we know there exists a set of params that satisfy the distance being 0. Unfortunately not the case for us
- Use an adaptive decreasing list. Using the quantile or median from the previous distances seems to be the generally accepted choice (e.g. it's used in https://pyabc.readthedocs.io/en/latest/index.html). The only problem is that ABC SMC can get stuck in a local minima if the global minima is too far to be attained by regular perturbations and enough randomness in the priors. For us it should be fine as I wouldn't expect wide multi-modal distributions and we have enough domain knowledge to set good priors